src/dataflow-types/src/types.rs

```diff
@@ -366,6 +366,39 @@ where
     }
 }
 
+impl<P, T> DataflowDescription<P, T> {
+    /// Determine a unique id for this dataflow based on the items it produces.
+    pub fn global_id(&self) -> Option<GlobalId> {
+        let mut desc_ids = self
+            .objects_to_build
+            .iter()
+            .map(|desc| desc.id)
+            .collect::<Vec<_>>();
+        desc_ids.sort_unstable();
+        desc_ids.dedup();
+        if desc_ids.len() == 1 {
+            return desc_ids.pop();
+        }
+        None
+    }
+}
+
+impl<P: PartialEq, T: timely::PartialOrder> DataflowDescription<P, T> {
+    pub fn compatible_with(&self, other: &Self) -> bool {
```

Comment:
For future work, but I'm happy to look at breaking `DataflowDescription` apart into a thing that can be equated (and which describes the collection we are building) and various instantiation arguments (e.g. `as_of`).

Reply:
#11445 so we don't forget.